crates/bevy_ui/src/widget/viewport.rs

```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    for (viewport_entity, pick_pointer_id) in viewport_picks {
+        let Ok((&viewport, &viewport_pointer_id, viewport_children)) =
+            viewport_query.get(viewport_entity)
+        else {
+            // This can only happen if entities in `dragged_last_frame` had one of these
+            // components removed since we last queried them
+            continue;
+        };
+
+        let Some((computed_node, image_node, global_transform)) = viewport_children
+            .iter()
+            .find_map(|child| node_query.get(*child).ok())
+        else {
+            continue;
+        };
+
+        let Some(cam_viewport_size) = camera_query
+            .get(viewport.camera)
+            .ok()
+            .and_then(Camera::logical_viewport_size)
+        else {
+            continue;
+        };
+
+        // Create a `Rect` in *physical* coordinates centered at the node's GlobalTransform
+        let node_rect = Rect::from_center_size(
+            global_transform.translation().truncate(),
+            computed_node.size(),
+        );
+        // Location::position uses *logical* coordinates
+        let top_left = node_rect.min * computed_node.inverse_scale_factor();
+        let logical_size = computed_node.size() * computed_node.inverse_scale_factor();
+
+        for input in pointer_inputs
+            .read()
+            .filter(|input| input.pointer_id == pick_pointer_id)
+        {
+            let local_position = (input.location.position - top_left) / logical_size;
+            let position = local_position * cam_viewport_size;
+
+            let location = Location {
+                position,
+                target: NormalizedRenderTarget::Image(image_node.image.clone().into()),
+            };
+
+            commands.send_event(PointerInput {
+                location,
+                pointer_id: viewport_pointer_id,
+                action: input.action,
+            });
+        }
+    }
+}
+
+/// Spawns a new viewport widget with the given `camera` and `target`.
+pub fn viewport(camera: Entity, target: Handle<Image>) -> impl Bundle {
+    (
+        Viewport::new(camera),
+        #[cfg(feature = "bevy_ui_picking_backend")]
+        PointerId::Custom(Uuid::new_v4()),
+        children![(
+            ImageNode::new(target),
+            Node {
+                position_type: PositionType::Absolute,
+                top: Val::ZERO,
+                bottom: Val::ZERO,
+                left: Val::ZERO,
+                right: Val::ZERO,
+                ..default()
+            },
+            #[cfg(feature = "bevy_ui_picking_backend")]
+            Pickable::IGNORE,
+        )],
+    )
+}
+
+/// Updates the size of the associated render target for viewports when the node size changes.
+pub fn update_viewport_render_target_size(
+    viewport_query: Query<(&Viewport, &Children)>,
+    node_query: Query<&ComputedNode, Changed<ComputedNode>>,
```

Comment:
Should watch for changes in the viewport camera too?

```suggestion
    viewport_query: Query<
        (&ViewportNode, &ComputedNode),
        (Changed<ComputedNode>, Changed<ViewportNode>),
    >,
```